### PR TITLE
feat: add GitHub Actions tools for workflow management

### LINKED ISSUE #268: Add support for Actions APIs
### Describe the feature or problem you’d like to solve

GitHub Actions has a suite of APIs that users may want to use with their AI-powered tools.

### Proposed solution

We add tools for Actions, starting with some of the core APIs like:

1. Listing workflows
2. Listing workflow runs
3. Triggering a workflow (via `workflow_dispatch`)
4. Re-running workflows
5. Downloading logs for a run

### Additional context

https://docs.github.com/rest/actions

### PR DESCRIPTION
**Pull request by gabornyergesX**
# Add GitHub Actions tools for workflow management

## Overview
This PR implements comprehensive GitHub Actions support for the MCP server, adding 14 new tools that enable complete workflow and CI/CD management through the GitHub Actions API.

## New Tools Added

https://github.com/user-attachments/assets/f1e29572-8115-4896-8421-17945209c58c

### Workflow Management (Read-only)
- **`list_workflows`** - List all workflows in a repository
- **`list_workflow_runs`** - List workflow runs for a specific workflow with filtering options
- **`get_workflow_run`** - Get detailed information about a specific workflow run
- **`get_workflow_run_logs`** - Download logs for a workflow run
- **`get_workflow_run_usage`** - Get usage metrics and billing information for workflow runs

### Job Management (Read-only)  
- **`list_workflow_jobs`** - List jobs for a specific workflow run
- **`get_job_logs`** - Download logs for a specific workflow job

### Artifact Management (Read-only)
- **`list_workflow_run_artifacts`** - List artifacts generated by a workflow run
- **`download_workflow_run_artifact`** - Get download URLs for workflow artifacts

### Workflow Operations (Write)
- **`run_workflow`** - Trigger workflows via workflow_dispatch events
- **`rerun_workflow_run`** - Re-run entire workflow runs
- **`rerun_failed_jobs`** - Re-run only failed jobs in a workflow run
- **`cancel_workflow_run`** - Cancel running workflows

### Administrative (Destructive)
- **`delete_workflow_run_logs`** - Delete workflow run logs (marked as destructive)

## Key Features

✅ **Complete CI/CD Workflow Support** - Covers the full lifecycle from triggering workflows to managing artifacts  
✅ **Proper Permission Classification** - Tools are correctly categorized as read-only, write, or destructive operations  
✅ **Comprehensive Parameter Validation** - All tools include proper input validation and error handling  
✅ **Rich Metadata** - Each tool includes detailed descriptions and parameter documentation  
✅ **Test Coverage** - Full test suite with 50+ test cases covering all scenarios  
✅ **Documentation** - Updated README with detailed tool descriptions and usage examples  

## Implementation Details

- **Architecture**: Follows existing MCP server patterns and conventions
- **Error Handling**: Consistent error responses with meaningful messages  
- **Pagination**: Supports GitHub API pagination for list operations
- **Filtering**: Includes advanced filtering options (by status, actor, branch, event, etc.)
- **Security**: Proper handling of GitHub tokens and permissions

## Testing

- Added comprehensive test suite in `actions_test.go`
- 50 test cases covering all tools and edge cases
- Mocked GitHub API responses for reliable testing
- All tests passing with proper error handling validation

## Breaking Changes

None - this is purely additive functionality.

## Closes

Closes #268 - Add support for Actions APIs

## Usage Example

```bash
# List workflows in a repository  
echo '{"tool":"list_workflows","arguments":{"owner":"github","repo":"github-mcp-server"}}' | github-mcp-server

# Trigger a workflow
echo '{"tool":"run_workflow","arguments":{"owner":"github","repo":"github-mcp-server","workflow_file":"ci.yml","ref":"main"}}' | github-mcp-server
```
```

### CODE REVIEW

**Comment by atxtechbro** [comment]
🚀 Excellent work @gabornyergesX! This is exactly what the community has been asking for.

I've tested this implementation in my fork and it's incredibly comprehensive:
- 14 new tools covering the complete GitHub Actions lifecycle
- Excellent test coverage with 50+ test cases
- Perfect integration with existing MCP server patterns
- Proper security annotations (read-only, write, destructive)
- Great documentation updates

This addresses issue #268 perfectly and goes above and beyond the original request. The code quality is exceptional - well done! 👏

Looking forward to seeing this merged so the whole community can benefit from these powerful CI/CD automation capabilities.

**Comment by SamMorrowDrums** [comment]
This looks very comprehensive. Thank you for putting this together. I hope I can review soon.

One preliminary question:

Is there strong motivation for allowing an agent to delete logs? I am hesitant to allow agents the ability to cover their tracks after executing something malicious. 

The security implications of this PR something I will need to consider. I really do want to land it though.

**Comment by atxtechbro** [comment]
## 💡 Enhancement Suggestion: Efficient Failed Job Logs

Great work on this comprehensive GitHub Actions implementation! After testing it, I noticed an opportunity to make it even more efficient for a common debugging use case.

### 🎯 Problem
Currently, to get logs for failed jobs, users must either:
1. Use `get_workflow_run_logs` - Downloads ALL logs (high token usage)
2. Manually combine `list_workflow_jobs` + `get_job_logs` - Multiple steps

This doesn't match the efficiency of `gh run view --log-failed`.

### 💡 Proposed Enhancement
Instead of adding a new tool, enhance the existing `get_job_logs` tool with optional parameters:

```go
// Enhanced parameters for get_job_logs
mcp.WithNumber("run_id", mcp.Description("Workflow run ID (required when using failed_only)")),
mcp.WithBoolean("failed_only", mcp.Description("When true, gets logs for all failed jobs in run_id")),
mcp.WithBoolean("return_content", mcp.Description("Returns actual log content instead of URLs")),
```

### 🔧 Usage Examples
```bash
# Original usage (unchanged)
get_job_logs --job_id=12345

# New: Get failed job logs efficiently  
get_job_logs --run_id=67890 --failed_only=true --return_content=true
```

### ✅ Benefits
- **Token efficient** - Only processes failed jobs
- **Content delivery** - Returns actual logs, not URLs (like `gh` CLI)
- **Backward compatible** - No breaking changes
- **Single tool** - No tool proliferation

This would make the Actions toolset perfect for real-world debugging workflows! 

I've created a detailed proposal in my fork: https://github.com/atxtechbro/github-mcp-server/blob/enhance/get-job-logs-failed-only/ENHANCEMENT_PROPOSAL.md

What do you think about this approach? Happy to help implement it if you're interested! 🚀

**Comment by gabornyergesX** [comment]
> This looks very comprehensive. Thank you for putting this together. I hope I can review soon.
> 
> One preliminary question:
> 
> Is there strong motivation for allowing an agent to delete logs? I am hesitant to allow agents the ability to cover their tracks after executing something malicious.
> 
> The security implications of this PR something I will need to consider. I really do want to land it though.

I did not have strong motivation for that just added as an extra.
My original issue was not able to debug pipeline logs with the agent and wanted to automise.
I can remove from the PR if you decide security implications are too high.

I think `delete_workflow_run_logs` tool has appropriate security measures:

1. ✅ **Properly marked as destructive** with `DestructiveHint: true`
2. ✅ **Full audit trail** via GitHub's native audit logging
3. ✅ **Requires proper permissions** (`actions:write` scope)
4. ✅ **Legitimate use cases** (storage management, compliance retention policies)

Let me know how to proceed. 
Thanks.

**Comment by gabornyergesX** [comment]
> ## 💡 Enhancement Suggestion: Efficient Failed Job Logs
> Great work on this comprehensive GitHub Actions implementation! After testing it, I noticed an opportunity to make it even more efficient for a common debugging use case.
> 
> ### 🎯 Problem
> Currently, to get logs for failed jobs, users must either:
> 
> 1. Use `get_workflow_run_logs` - Downloads ALL logs (high token usage)
> 2. Manually combine `list_workflow_jobs` + `get_job_logs` - Multiple steps
> 
> This doesn't match the efficiency of `gh run view --log-failed`.
> 
> ### 💡 Proposed Enhancement
> Instead of adding a new tool, enhance the existing `get_job_logs` tool with optional parameters:
> 
> ```go
> // Enhanced parameters for get_job_logs
> mcp.WithNumber("run_id", mcp.Description("Workflow run ID (required when using failed_only)")),
> mcp.WithBoolean("failed_only", mcp.Description("When true, gets logs for all failed jobs in run_id")),
> mcp.WithBoolean("return_content", mcp.Description("Returns actual log content instead of URLs")),
> ```
> 
> ### 🔧 Usage Examples
> ```shell
> # Original usage (unchanged)
> get_job_logs --job_id=12345
> 
> # New: Get failed job logs efficiently  
> get_job_logs --run_id=67890 --failed_only=true --return_content=true
> ```
> 
> ### ✅ Benefits
> * **Token efficient** - Only processes failed jobs
> * **Content delivery** - Returns actual logs, not URLs (like `gh` CLI)
> * **Backward compatible** - No breaking changes
> * **Single tool** - No tool proliferation
> 
> This would make the Actions toolset perfect for real-world debugging workflows!
> 
> I've created a detailed proposal in my fork: https://github.com/atxtechbro/github-mcp-server/blob/enhance/get-job-logs-failed-only/ENHANCEMENT_PROPOSAL.md
> 
> What do you think about this approach? Happy to help implement it if you're interested! 🚀

should i remove get_workflow_run_logs completly?

https://github.com/github/github-mcp-server/pull/491/commits/15ca4b7e3371d6fef73b77676073f8af0c1abee3

**Comment by atxtechbro** [comment]
## 🎉 Excellent Implementation! 

@gabornyergesX - I just reviewed your latest commit and you've implemented **exactly** what we discussed! This is fantastic work.

### ✅ Perfect Implementation

Your commit shows you've added:
- ✅ **`io` and `net/http` imports** - exactly what was needed
- ✅ **`failed_only` parameter** for efficient failed job filtering  
- ✅ **`return_content` parameter** for actual log content vs URLs
- ✅ **`downloadLogContent()` function** - downloads actual logs like `gh` CLI
- ✅ **Comprehensive test coverage** - 319 new test lines!
- ✅ **Backward compatibility** - existing usage unchanged

### 🎯 Answering Your Question

**Should you remove `get_workflow_run_logs` completely?**

**No, keep it!** Here's why:

1. **Different use cases**:
   - `get_workflow_run_logs` - Complete workflow archive (ZIP) for comprehensive analysis
   - `get_job_logs` with `failed_only=true` - Targeted debugging (your new enhancement)

2. **Your warnings are perfect**:
   ```go
   "EXPENSIVE: downloads ALL logs as ZIP. Consider using get_job_logs with failed_only=true for debugging failed jobs"
   ```

3. **Complementary tools**:
   - Keep both with clear guidance on when to use each
   - Your optimization tips guide users to the efficient option

### 🚀 This is Production Ready!

Your implementation:
- ✅ **Solves the token efficiency problem** - only downloads failed job logs
- ✅ **Matches `gh run view --log-failed`** - returns actual content
- ✅ **Maintains backward compatibility** - no breaking changes  
- ✅ **Excellent error handling** - comprehensive validation
- ✅ **Great UX** - clear warnings and optimization tips

This is exactly what debugging workflows needed! The combination of both tools gives users flexibility while guiding them toward the efficient approach.

**Recommendation: Keep both tools as-is** - your implementation is perfect! 🎯

**Comment by atxtechbro** [comment]
## 🔧 Build Issue - Ready to Test When Fixed

@gabornyergesX - Excited to test your enhanced `get_job_logs` implementation! 

Just pulled your latest changes but getting some build errors:

```
pkg/github/actions.go:22:19: undefined: toBoolPtr
pkg/github/actions.go:40:18: undefined: requiredParam
pkg/github/actions.go:44:17: undefined: requiredParam
```

Looks like some helper functions might be missing or need to be imported differently.

**Let us know when it's ready to test!** We have a current debugging need and would love to try out the new `failed_only=true` + `return_content=true` functionality. 🚀

Thanks for the great work on this!

**Comment by SamMorrowDrums** [comment]
Those helper functions were just renamed to export them. So just need capital letter.

**Comment by gabornyergesX** [comment]
> ## 🔧 Build Issue - Ready to Test When Fixed
> @gabornyergesX - Excited to test your enhanced `get_job_logs` implementation!
> 
> Just pulled your latest changes but getting some build errors:
> 
> ```
> pkg/github/actions.go:22:19: undefined: toBoolPtr
> pkg/github/actions.go:40:18: undefined: requiredParam
> pkg/github/actions.go:44:17: undefined: requiredParam
> ```
> 
> Looks like some helper functions might be missing or need to be imported differently.
> 
> **Let us know when it's ready to test!** We have a current debugging need and would love to try out the new `failed_only=true` + `return_content=true` functionality. 🚀
> 
> Thanks for the great work on this!

Sorry @atxtechbro @SamMorrowDrums ,
there was an issue on my side, i rebased to main.
Should compile now, tested locally.

**Comment by tonytrg** [comment]
@gabornyergesX could you fix the merge request and we can get to review it 🥳 very great pr and contribution

**Comment by atxtechbro** [comment]
## ✅ Confirmed: Ready to Test!

@gabornyergesX - Just pulled your latest changes and confirmed:

✅ **Builds successfully** - no more `toBoolPtr`/`requiredParam` errors  
✅ **Enhanced `get_job_logs` working** - with `failed_only=true` + `return_content=true`  
✅ **All 14 Actions tools functional**

Your implementation is **production ready**! The enhanced failed job logs feature works exactly like `gh run view --log-failed` - super efficient for debugging.

@tonytrg mentioned fixing merge conflicts - happy to help with that if needed, but the core functionality is solid and ready for maintainer review! 🚀

**Comment by tonytrg** [comment]
nit: could we get an updated toolset section in here for actions. https://github.com/github/github-mcp-server/blob/main/README.md?plain=1#L193

The docs are maintained by hand currently and it is easy to overlook what needs to be updated, that is something we need to fix in the future.

**Comment by gabornyergesX** [comment]
> nit: could we get an updated toolset section in here for actions. https://github.com/github/github-mcp-server/blob/main/README.md?plain=1#L193
> 
> The docs are maintained by hand currently and it is easy to overlook what needs to be updated, that is something we need to fix in the future.

@tonytrg let me know if there are some missing gaps.

File: pkg/github/actions.go [hunk 1]
```diff
@@ -0,0 +1,1156 @@
...
+			mcp.WithString("actor",
+				mcp.Description("Returns someone's workflow runs. Use the login for the user who created the workflow run."),
+			),
+			mcp.WithString("branch",
+				mcp.Description("Returns workflow runs associated with a branch. Use the name of the branch."),
+			),
+			mcp.WithString("event",
+				mcp.Description("Returns workflow runs for an event. For example, push, pull_request, or issue."),
+			),
+			mcp.WithString("status",
```
Comment (tonytrg) [inline pkg/github/actions.go:155]: All strings that have enum types should be marked so like here:
https://github.com/github/github-mcp-server/blob/main/pkg/github/code_scanning.go#L93-L97

Examples should then be removed from the description: meaning `example, completed, in_progress, or requested.`

I am not sure if I marked all occurences of params that should be converted to enums, but we should update those strings with their enum configuration.

File: pkg/github/actions.go [hunk 2]
```diff
@@ -0,0 +1,1156 @@
...
+				mcp.Description("The account owner of the repository. The name is not case sensitive."),
+			),
+			mcp.WithString("repo",
+				mcp.Required(),
+				mcp.Description("Repository name"),
+			),
+			mcp.WithString("workflow_file",
+				mcp.Required(),
+				mcp.Description("The workflow ID or workflow file name"),
+			),
```
Comment (tonytrg) [inline pkg/github/actions.go:259]: we are labelling this input param as id or workflow, but the func impl is only able to handle `CreateWorkflowDispatchEventByFileName`/filenames?

File: pkg/github/actions.go [hunk 3]
```diff
@@ -0,0 +1,1156 @@
...
+			mcp.WithString("repo",
+				mcp.Required(),
+				mcp.Description("Repository name"),
+			),
+			mcp.WithNumber("run_id",
+				mcp.Required(),
+				mcp.Description("The unique identifier of the workflow run"),
+			),
+			mcp.WithString("filter",
```
Comment (tonytrg) [inline pkg/github/actions.go:481]: Here we should turn it into enum and remove enums from description

File: pkg/github/actions.go [hunk 4]
```diff
@@ -0,0 +1,1156 @@
...
+// RunWorkflow creates a tool to run an Actions workflow
+func RunWorkflow(getClient GetClientFn, t translations.TranslationHelperFunc) (tool mcp.Tool, handler server.ToolHandlerFunc) {
+	return mcp.NewTool("run_workflow",
+			mcp.WithDescription(t("TOOL_RUN_WORKFLOW_DESCRIPTION", "Run an Actions workflow")),
+			mcp.WithToolAnnotation(mcp.ToolAnnotation{
+				ReadOnlyHint: ToBoolPtr(false),
+			}),
+			mcp.WithString("owner",
+				mcp.Required(),
+				mcp.Description("The account owner of the repository. The name is not case sensitive."),
```
Comment (tonytrg) [inline pkg/github/actions.go]: Could we replace all occurences `The account owner of the repository. The name is not case sensitive.` with `Repository owner`

File: pkg/github/actions.go [hunk 5]
```diff
@@ -0,0 +1,1156 @@
...
+		}
+}
+
+// ListWorkflowRuns creates a tool to list workflow runs for a specific workflow
+func ListWorkflowRuns(getClient GetClientFn, t translations.TranslationHelperFunc) (tool mcp.Tool, handler server.ToolHandlerFunc) {
+	return mcp.NewTool("list_workflow_runs",
+			mcp.WithDescription(t("TOOL_LIST_WORKFLOW_RUNS_DESCRIPTION", "List workflow runs for a specific workflow")),
+			mcp.WithToolAnnotation(mcp.ToolAnnotation{
+				ReadOnlyHint: ToBoolPtr(true),
+			}),
```
Comment (tonytrg) [inline pkg/github/actions.go:99]: Image

Can we get some short precise annotation titles here for a nicer UI experience.

**Review by tonytrg — COMMENTED** [review]
Didn't manage to finish the review today, but some early feedback. Great work!

**Comment by gabornyergesX** [comment]
> Didn't manage to finish the review today, but some early feedback. Great work!

i will review your suggestions and will fix. @tonytrg

File: pkg/github/actions.go (hunk 3 above)
Comment (gabornyergesX) [reply pkg/github/actions.go:481]: ✅ **Done!** Converted the `filter` parameter to use enum with `["latest", "all"]` values and removed the example text from the description.

File: pkg/github/actions.go (hunk 1 above)
Comment (gabornyergesX) [reply pkg/github/actions.go:155]: **✅ Done!** Converted all enum parameters in actions.go:

- **`status`** → `mcp.Enum("queued", "in_progress", "completed", "requested", "waiting")`
- **`filter`** → `mcp.Enum("latest", "all")`  
- **`event`** → `mcp.Enum()` with all 32 Free/Pro/Team events from [GitHub docs](https://docs.github.com/en/actions/writing-workflows/choosing-when-your-workflow-runs/events-that-trigger-workflows)

Removed all example text from descriptions as requested. The enums now match the official GitHub API documentation exactly.

File: pkg/github/actions.go (hunk 2 above)
Comment (gabornyergesX) [reply pkg/github/actions.go:259]: **✅ Fixed!** Implemented the **go-github pattern** with two separate functions:

- **`run_workflow`** → workflow ID (numeric) 
- **`run_workflow_by_filename`** → workflow filename (string)

This follows the exact same approach as [go-github's solution](https://github.com/google/go-github/issues/1639) where they provide `CreateWorkflowDispatchEvent` and `CreateWorkflowDispatchEventByFileName` as separate methods.

**Result**: Parameter descriptions now accurately match implementations - no more misleading "workflow ID or filename" descriptions.

File: pkg/github/actions.go (hunk 4 above)
Comment (gabornyergesX) [reply pkg/github/actions.go]: ✅ **Refactored to DRY constants for parameter descriptions**

Extracted repeated string literals into constants to improve maintainability:
- Added `DescriptionRepositoryOwner` and `DescriptionRepositoryName` constants
- Replaced 30+ hardcoded occurrences across all GitHub Actions tools

File: pkg/github/actions.go (hunk 5 above)
Comment (gabornyergesX) [reply pkg/github/actions.go:99]: ✅ **Added user-friendly titles to all GitHub Actions tools**

Following the `notifications.go` pattern, added `Title` annotations to all 15 Actions tools for better UI experience in MCP clients. Now tools display clear names like "List workflows" and "Run workflow" instead of technical function names.

Comment (tonytrg) [reply pkg/github/actions.go:99]: @gabornyergesX did you push the latest changes?

File: pkg/github/actions.go (hunk 2 above)
Comment (tonytrg) [reply pkg/github/actions.go:259]: Sorry If i was not clear, RunWorkflow and RunWorkflowByFileName are now almost duplicates.

I meant for the `workflow_id Required` to be like in the [api]:(https://docs.github.com/en/rest/actions/workflows?apiVersion=2022-11-28#create-a-workflow-dispatch-event)
The ID of the workflow. You can also pass the workflow file name as a string.

So imo:
- a single `RunWorkflow` function
- workflow_id as string
- casting check to see if its an id
  - if its int `client.Actions.CreateWorkflowDispatchEventByID(ctx, owner, repo, workflowID, event)`
  - if its string `client.Actions.CreateWorkflowDispatchEventByFileName(ctx, owner, repo, workflowIDasFilename, event)`

Comment (gabornyergesX) [reply pkg/github/actions.go:259]: thanks for clarification!

File: pkg/github/actions.go (hunk 5 above)
Comment (gabornyergesX) [reply pkg/github/actions.go:99]: now https://github.com/github/github-mcp-server/pull/491/commits/f7e1320af7f64a7ee6440d07ac0359eaa484b892

**Review by tonytrg — APPROVED** [review]
Excellent work @gabornyergesX and @atxtechbro 

Very much appreciated, it works very well.

**Comment by gabornyergesX** [comment]
> Excellent work @gabornyergesX and @atxtechbro
> 
> Very much appreciated, it works very well. One more time to appease the linter.

➜  github-mcp-server git:(main) ✗ golangci-lint run                            
0 issues.

**Review by tonytrg — APPROVED** [review]

**Comment by tonytrg** [comment]
Great stuff!

**Comment by gabornyergesX** [comment]
> Great stuff!

It was a pleasure to work with you @atxtechbro , @tonytrg

**Comment by Mostafa-wael** [comment]
How to get these tools if I am using GitHub host?
I know that for docker, I can use `GITHUB_TOOLSETS="repos,issues,pull_requests,actions,code_security,experiments" `
But, when I am using GitHub host like this:
```
"github": {
        "type": "http",
        "url": "https://api.githubcopilot.com/mcp/",
        "headers": {
            "Authorization": "Bearer ${input:github_token}"
        }
        "env": {
            "GITHUB_TOOLSETS": "all"
        }
      }
```
I am getting `Property env is not allowed.`

**Comment by SamMorrowDrums** [comment]
@Mostafa-wael hi, so the docs are in this repo here: doc/remote-server.md

But this feature hasn't been released yet, and is only on the head of main. We wanted to improve a couple of things but we will release it in a semver build of this and in the remote server this week.
